src/System.Net.Security/src/System/Net/SecureProtocols/_FixedSizeReader.cs

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.IO;
+
+namespace System.Net
+{
+    //
+    // Reads a fixed size packet from a stream, can disover EOF as 0 bytes read from stream.
+    //
+    internal class FixedSizeReader
+    {
+        private static readonly AsyncCallback s_readCallback = new AsyncCallback(ReadCallback);
+
+        // TODO (Issue #3114): Implement this using TPL instead of APM.
+        private readonly StreamAsyncHelper _TransportAPM;
+        private readonly Stream _Transport;
+        private AsyncProtocolRequest _Request;
+        private int _TotalRead;
+
+        public FixedSizeReader(Stream transport)
+        {
+            _Transport = transport;
+            _TransportAPM = new StreamAsyncHelper(transport);
+        }
+
+        //
+        // Returns 0 on legitimate EOF or if 0 bytes was requested, otheriwse reads as directed or throws.
+        // Returns count on success.
+        //
+        public int ReadPacket(byte[] buffer, int offset, int count)
+        {
+            int tempCount = count;
+            do
+            {
+                int bytes = _Transport.Read(buffer, offset, tempCount);
+
+                if (bytes == 0)
+                {
+                    if (tempCount != count)
```

Comment:
I don't quite follow this.  It feels like asking for 1024 bytes when 12 remain would do a loop that read 12, then throw.  Is that right? Is that good?  What happened to the last 12 bytes? (Oh, won't somebody think of the 12 bytes?!)

Some comments might be warranted :smile:.